### ISSUE TITLE
Using nagstamon with a Thruk installation that's authenticated using Azure AD - is this possible?

### ISSUE THREAD

**Issue by ianbamforth** [HenriWahl/Nagstamon#806]
We have a thruk installation that's got Azure AD in front of it - does this make it impossible for nagstamon to talk to it, or is there  a workaround?

**Comment by HenriWahl** [comment]
What do you mean by "in front of it"? Another screen aside the Thruk one? If so, I guess it won't work.

**Comment by pvdputte** [comment]
I suppose the OP is using the Oauth support in Thruk. See [https://www.thruk.org/documentation/oauth2.html](https://www.thruk.org/documentation/oauth2.html)

I'm using it with Okta, not AD, but the idea is the same. To log in on Thruk, you are redirected to the authentication provider, who redirects back to Thruk after a successful login, with a bunch of authentication tokens.

AFAICT Nagstamon polls Thruk's status.cgi, requesting JSON output. Couldn't this work with [the Thruk API](https://www.thruk.org/documentation/rest.html#_endpoints) instead, and use Thruk's API tokens? Each Thruk user that wants to use Nagstamon can then create an API token by himself and configure Nagstamon with it, instead of relying on basic auth with username/password.

Making authenticated REST requests (and [filtering them](https://www.thruk.org/documentation/rest.html#_complex-filtering)) is then very simple, by using an `X-Thruk-Auth-Key: your_key_here` header. Example:

```
$ curl -s -H 'X-Thruk-Auth-Key: e41deadbeefoodbabe2323424327d_1' -d 'q=***state != 0***' https://your.thruk.fqdn/thruk/r/services | jq '.[] | {host_name,description,plugin_output}'
{
  "host_name": "some.host.net",
  "description": "LINK-STATE-1234",
  "plugin_output": "CRITICAL: Interface ae46 on 1.2.3.4 is down (operational state is 2)"
}
```

**Comment by pvdputte** [comment]
Hmm 🤔 

Or is that what "autologin" does? It refers to `X-Thruk-Auth-Key`.

https://github.com/HenriWahl/Nagstamon/commit/38a306fbdba63eccf58a64352f070efa6fbabe9d

**Comment by HenriWahl** [comment]
As these ways for login are increasingly used I wish there would be a way, I have the same problem with Checkmk. Currently there is no solution. I fear this would need a custom solution for every single server type.

**Comment by HenriWahl** [comment]
@pvdputte the issue https://github.com/HenriWahl/Nagstamon/issues/1125 inspired me to try to add browser based login. I am experimenting with the QWebEngine part of Qt. This might be the solution for your issue too.

**Comment by pvdputte** [comment]
Thanks, I'll definitely keep an eye on this. 👍

**Comment by HenriWahl** [comment]
@ianbamforth @pvdputte there is a pretty raw POC now available as latest testing release,  maybe it already works for you. Right now the authentication mode has to be Web and the username and password empty.

**Comment by AimbotNooby** [comment]
Hey @HenriWahl, where can I get the testing release? I would like to test authentication with Entra-ID.

**Comment by HenriWahl** [comment]
Get it at https://github.com/HenriWahl/Nagstamon/releases/tag/latest

Just set authentication to "Web".

**Comment by pvdputte** [comment]
FYI, I was curious what would happen so I tried the latest release 3.17-20250911, setting authentication to 'Web' and leaving username/password empty.

Nothing really happens after that, in the console I see this message passing by.

```
DEBUG: 2025-09-15 16:04:36.116037 somefqdn Login status: https://somefqdn/thruk/cgi-bin/login.cgi?/thruk/ http code : 401
```

So I guess it fetches the Thruk login page with the "[Login with okta](https://www.thruk.org/documentation/oauth2.html#_okta)" button (that redirects you to the auth provider), but that's it.

TBH every other day we need to re-authenticate with Okta, so even if it would work, it would not be that practical. I feel (for Thruk at least) calling the Thruk api with an apikey would be the simplest. I understand you don't feel like maintaining custom solutions for every server type. Thanks for the effort 🙂

**Comment by HenriWahl** [comment]
@pvdputte actually another little browser window should open, allowing to use the credentials for whatever identity solution and whatever forwards. It therefrom will get an authorized cookie and use it itself without browser. So far the theory. Sounds like there is no extra little window in your case?

**Comment by pvdputte** [comment]
During my testing yesterday I had to re-run the executable multiple times before I actually got the main UI. (perhaps wayland-related, running Arch here).

So I tried again and indeed, after a few times I got the web login in an extra little window and, it works indeed! 👍 
No idea what will happen though after a few days when I need to re-authenticate. I suppose the 'Authentication problem' warning will then flash orange again.

Anyway, I'm afraid I have to report the next issue with this approach, when adding a second Thruk backend this way: 😬 
```
ERROR: 2025-09-16 11:32:30.632665 INSTANCE2 Traceback (most recent call last):
  File "Nagstamon/servers/Thruk.py", line 82, in init_http
  File "requests/cookies.py", line 202, in get
  File "requests/cookies.py", line 405, in _find_no_duplicates
requests.cookies.CookieConflictError: There are multiple cookies with name, 'thruk_auth'
```
Sorry... 🙂

**Comment by HenriWahl** [comment]
I feared this day will come... of course, as long as Nagstamon throws all cookies into one store this must happen... well, I will check this out. Still under construction anyway - so thanks for finding problems.

**Comment by HenriWahl** [comment]
@pvdputte did you add a second login on the same server with the same user? Or are there any differences?

**Comment by pvdputte** [comment]
No, it's a fully separate Thruk backend, but set up in an identical way (same software version, authenticating to the same okta backend).

**Comment by HenriWahl** [comment]
@pvdputte please check the latest testing release which better differences cookies.

**Comment by pvdputte** [comment]
Yes, it's working fine now 👌 

And after a few days the "Authentication problem" button appears and I just need to do the web login again. I'm not sure if it remembers the username etc in the pop-up window.. But it works👍 

The only thing left I'm running into is a CSRF token issue when sending commands to Thruk, but apparently that's well-known and a workaround is available. https://www.thruk.org/documentation/faq.html#_possible-csrf-no-or-invalid-token

Thanks for the great work.
